src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,297 @@
+
+          Date value = (Date) entry.getValue();
+          formattedValue = formatDate(value);
+
+        } else if (entry.getValue() instanceof String && ISO_DATE.matcher(String.valueOf(entry.getValue())).matches()) {
+
+          formattedValue = String.valueOf(entry.getValue()).substring(0, 10);
+
+        } else {
 
+          formattedValue = String.valueOf(entry.getValue());
+        }
+
+        mFirebaseAnalytics.setUserProperty(trait, formattedValue);
+        logger.verbose("mFirebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+      }
+    }
   }
 
-  @Override public void screen(ScreenPayload screen) {
+  @Override
+  public void screen(ScreenPayload screen) {
     super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+
+    logger.verbose(
+            "The Firebase SDK gathers screen calls automatically. Segment does map manual screen" +
+                    "methods to Firebase.");
+
   }
 
-  @Override public void track(TrackPayload track) {
+  @Override
+  public void track(TrackPayload track) {
     super.track(track);
 
+    String event = track.event();
+    String eventName = mapEvent(event);
+
+    Properties properties = track.properties();
+    Bundle formattedProperties = formatProperties(properties);
+
+    mFirebaseAnalytics.logEvent(eventName, formattedProperties);
+    logger.verbose("mFirebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
+  }
+
+  private String mapEvent(String event) {
+
+    String eventName = event;
+
+    final Map<String, String> eventMapper = new HashMap<>();
+    eventMapper.put("Product Added", Event.ADD_TO_CART);
+    eventMapper.put("Checkout Started", Event.BEGIN_CHECKOUT);
+    eventMapper.put("Order Completed", Event.ECOMMERCE_PURCHASE);
+    eventMapper.put("Order Refunded", Event.PURCHASE_REFUND);
+    eventMapper.put("Product Viewed", Event.VIEW_ITEM);
+    eventMapper.put("Product List Viewed", Event.VIEW_ITEM_LIST);
+    eventMapper.put("Payment Info Entered", Event.ADD_PAYMENT_INFO);
+    eventMapper.put("Promotion Viewed", Event.PRESENT_OFFER);
+    eventMapper.put("Product Added to Wishlist", Event.ADD_TO_WISHLIST);
+    eventMapper.put("Product Shared", Event.SHARE);
+    eventMapper.put("Product Clicked", Event.SELECT_CONTENT);
+    eventMapper.put("Product Searched", Event.SEARCH);
+    eventMapper.put("Promotion Viewed", Event.PRESENT_OFFER);
+
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
+    }
+
+    if (eventName.contains(" ")) {
+      eventName = eventName.trim().toLowerCase().replaceAll(" ", "_");
+    }
+
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+
+    return eventName;
+  }
+
+  private Bundle formatProperties(Properties properties) {
+
+    if ((properties.revenue() != 0 || properties.total() != 0) && isNullOrEmpty(properties.currency())) {
+      logger.verbose(
```

Comment:
This is just logging - should it not bail if this "must" be set?

Reply:
Added try/catch block starting on line 187 and threw an exception on line 214 if a customer doesn't pass currency. Let me know if this was the right approach. Thanks!

```
    try {
      Bundle formattedProperties = formatProperties(properties);
      firebaseAnalytics.logEvent(eventName, formattedProperties);
      logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
    } catch (Exception e) {
      logger.verbose(
              "You must set `currency` in your event's property object to accurately "
                      + "pass 'value' to Firebase. This event was not sent to Firebase.");
    }
```

Reply:
Sorry that's not quite what I meant. What  I was asking was that instead of only logging, shouldn't you be calling `return` or `break` to stop further execution if this property "must" be sent? Otherwise it's just an optional property and probably shouldn't be logged.

In this case however, I don't think we need logging or bailing. We should instead just have a fallback. If currency is not set, it should be US dollars https://segment.com/docs/spec/track/.

> This should be sent in the ISO 4127 format. If this is not set, we assume the revenue to be in US dollars.

Lastly, using exceptions for control flow is not a good practice (in java and most other languages).